### DISCUSSION TITLE
MBP trackpad's force click setting causes undesired behavior

### DISCUSSION THREAD

**Discussion by rcdailey** [rxhanson/Middle-Localization#88] [General]
I have a new 2023 MBP. I ended up turning off "Force click and haptic feedback" option under Trackpad Settings, because I found that it was too easy to accidentally push too hard and cause that secondary click. Unfortunately, when this happens, it registers as two separate clicks. So for example in the MS Edge Browser, when I use three-finger clicks to close tabs and the secondary/force click happens, it closes 2 tabs instead of one.

Even though I have that force click setting set to OFF, sometimes that secondary click still happens! I don't know if this is a bug with Middle or MacOS. I think in either case, it would be reasonable to request that there be a setting in Middle that allows me to ignore the secondary/force click. In other words, even if that force click *does* happen, enabling this new setting would prevent that from being treated as yet-another middle mouse button click.

**Comment by rxhanson** [comment]
Thanks for reporting. Unfortunately, as far as I know there is no way to prevent that additional click. The reason is because macOS provides no way to differentiate clicks. It's been a while since I've looked into this, though, so I'll see if I can find any different results with a fresh perspective.

**Comment by rcdailey** [comment]
Could you implement a customizable delay? For example, any middle click that happens within 100ms of the previous click gets dropped? I think this is a reasonable workaround because it doesn't require disambiguating the click types, and also I think personally it's hard to be humanly able to click again within 100ms, but hey that's why it's customizable I guess?

So for the UI, I would recommend:

1. A "Click Debounce" setting with both an edit field to configure the delay, as well as a checkbox to enable/disable the feature entirely.
2. When the checkbox is enabled, it allows an edit field to be modified which contains the debounce delay. It would default to some reasonable delay, such as 100ms.
3. When the checkbox is disabled, the edit box is either hidden or remains shown but not editable.
4. If a value of 0ms is entered, this fails input validation. If this should be considered valid input, alternatively to disable the feature 0ms can be used instead of a checkbox. But I do not feel this is good UX.
5. This setting would default to OFF to avoid impacting existing users.

Hope that seems reasonable, assuming you don't find any better solutions!

**Comment by nullsubject** [comment]
I don't know if I am experiencing the same issue, or a slightly different one. I'm on an M4 MBP, and occasionally my clicks are doubled up. On my end, it's very clear that it isn't the force click activating, because I can press harder and get that force click to activate, and it feels quite different (I'm also not pressing nearly hard enough to force click). This is reproducible on my end just by three finger clicking around, at which point some clicks will be registered as instant double click (I do feel the haptics twice in extremely quick succession, and two tabs are opened).

I have no reason to believe that my trackpad is faulty as it works as expected otherwise, and never double clicks on left or right clicks.

The user suggested solution above seems like it would be perfect in my case, and in his, but this thread is quite old by now. Are there any plans to fix this issue?

Thank you very much for the hard work, besides this annoyance, I've enjoyed your app tremendously.

**Reply by rxhanson** [reply]
In Middle, do you have it set to the three finger click as well? is there anything unusual about your setup, like having tap-to-click enabled in macOS?

**Reply by nullsubject** [reply]
I don't believe there is anything unusual about my setup, and this is a brand new laptop (no app conflict/clutter).

I don't use tap to click, and everything in middle is set to three finger clicks. I do believe a debounce would fix my issue (it's hard to believe that not everyone experiences this?). I could record a video of the issue if you think it necessary. This is literally the only niggle I see with your app. Thanks for the quick reply, I really appreciate it!

<img width="622" alt="Screenshot 2025-03-07 at 3 43 54 AM" src="https://github.com/user-attachments/assets/7afb56a6-e70b-4303-a2be-59517f3de7c8" />
<img width="827" alt="Screenshot 2025-03-07 at 3 44 17 AM" src="https://github.com/user-attachments/assets/d941bd3c-fc5b-46d4-ba64-c35eef57ac9a" />

**Comment by rxhanson** [comment]
Apologies for not responding to this in quite some time. I can see the appeal for at least having some sort of debounce in place for scenarios like this and will take another look.

**Reply by nullsubject** [reply]
As a new user experiencing this singular issue, this is absolutely wonderful news! Thank you very much!

**Comment by nullsubject** [comment]
I've just seen the update for middle, and that it fixed the issue with magic mice. I can confirm my mouse seems to work as expected now, even after being turned off and on. This debounce functionality is the only issue left with most/all middle click software i've used on mac. Obviously, this may take quite a bit of work on your end, so it's not my intention to push you to work faster, but I did want to make sure you hadn't forgotten about this, and/or that you still considered adding such a functionality to middle. Is that still in the cards? 

Thank you, and have a good day!

**Reply by rxhanson** [reply]
Thanks for checking back in! Apologies for not following up. I still don't mind building in something for this, but one thing that prevented me from prioritizing it for this recent release is that it actually does sound like a hardware issue (or software other than Middle) to me. I would almost never think down this path because of Apple hardware quality being so good. Middle won't trigger any haptics, and I had an old MacBook Pro that would also sometimes trigger haptics twice with three finger clicks. It wasn't as directly reproducible as your issue sounds, though (and I no longer have that Mac). A fix can be added to Middle to debounce here, although you would still get the double haptics since that's not related to Middle.

**Reply by nullsubject** [reply]
Hmmm, I'm not sure what you mean by "middle doesn't trigger any haptics." Are you saying that there should be no haptic feedback when three finger clicking? Not tapping: clicking. Every single app besides BetterTouchTool that I've tried has had the same click haptic as single, or two finger clicking (default mac haptic). Better touch tool has configurable haptics, so by default the three finger click "feels" very different to single or two finger click, but nonetheless it too has some feedback. Just to be clear, the double haptic appears to be caused by the actual doubled input, not middle adding a random haptic impulse or anything. I very much doubt that the double haptic would still happen if the double input wasn't registered, since it seems like the haptic is cause by the input and not the reverse. Obviously, you probably know much better than I do.

I suppose it could be a hardware issue, but then again, multiple people are experiencing this, and I can't easily think of a reason why _only_ middle clicking would be an issue. Never had a single finger doubled click, or two finger double click, never had a gesture fail to trigger, etc etc. Stranger things have happened, obviously, but I still have some doubt.

Could a video of the issue be of any help?

**Reply by rxhanson** [reply]
Thanks for asking. 

My expectation is that if you were to quit Middle and use three fingers to left click everywhere instead of one, that you would experience the duplicate clicks. That was my experience with the MacBook Pro that I had a long time ago, although it was fairly intermittent. If you have a separate trackpad to test, then I would expect it to not have this issue. I could be wrong though; there are a lot of things that can affect input events. I think BetterTouchTool might work differently by overriding the macOS click behavior entirely and providing its own, and if the issue is at the hardware level, then it would also need to compensate for the duplicate clicks. Your situation is indeed the only report of this issue that I've come across, as the initial issue here was with Force Touch and is different. All that said, I'd like to see if I can get something that would work but it might take some time.
